common/kotlinx-coroutines-core-common/src/internal/ArrayQueue.kt

```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright 2016-2018 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package kotlinx.coroutines.internal
+
+internal class ArrayQueue<T : Any> {
+    public val isEmpty: Boolean get() = head == tail
+
+    private var elements = arrayOfNulls<Any>(16)
+    private var head = 0
+    private var tail = 0
+
+    public fun addLast(element: T) {
+        elements[tail] = element
+        tail = (tail + 1) and elements.size - 1
+        if (tail == head) ensureCapacity()
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    public fun removeFirst(): T {
+        require(head != tail) { "Queue is empty" }
+        val element = elements[head]
+        elements[head] = null
+        head = (head + 1) and elements.size - 1
+        return element!! as T
```

Comment:
Consider dropping `<T: Any>` requirement and (leave just `<T>`). It would remove the need to have `!!` null-check here.